### PR TITLE
[api-minor] Simplify API to implement zoom in custom viewers

### LINKED ISSUE #18076: Feature request: Improve scaling API to simplify zoom implementation in embedders
To properly implement zoom gestures (Ctrl+wheel, or pinch-to-zoom), applications that embed the viewer currently have to:
1. compute the scaling factor/steps based on the input method
2. call `pdfViewer.decreaseScale` or `pdfViewer.increaseScale`, depending on whether it's a zoom in or out
3. scroll the PDF to restore the center of the gesture, knowing that `decreaseScale`/`increaseScale` preserve the position of the top-left visible corner.

<details>
<summary>As an example, in the PDF.js app, (2) needs to be checked 4 times (one for each zoom input method):</summary>

https://github.com/mozilla/pdf.js/blob/b6765403a1b08aa80a21cb0809741875a925ab4d/web/app.js#L2641-L2646
https://github.com/mozilla/pdf.js/blob/b6765403a1b08aa80a21cb0809741875a925ab4d/web/app.js#L2679-L2685
https://github.com/mozilla/pdf.js/blob/b6765403a1b08aa80a21cb0809741875a925ab4d/web/app.js#L2800-L2806
https://github.com/mozilla/pdf.js/blob/b6765403a1b08aa80a21cb0809741875a925ab4d/web/app.js#L2813-L2819

</details>

And for (3) it has a "post-zoom scroll fixup" implemented at https://github.com/mozilla/pdf.js/blob/b6765403a1b08aa80a21cb0809741875a925ab4d/web/app.js#L2101-L2109

The same logic needs to be re-implemented in other embedders that implement gestures-based zoom, such as [in ngx-extended-pdf-viewer](https://github.com/stephanrauh/ngx-extended-pdf-viewer/blob/03c9bfce672a60f0ebe5edacc160a23c5cbba753/projects/ngx-extended-pdf-viewer/src/lib/pinch-on-mobile-support.ts#L99-L103) for angular (or the internal one I'm helping with, which re-implements similar logic).

---

What I would like the viewer to provide is:
- [ ] a `center: [x, y]` parameter in `increaseScale`/`decreaseScale` that allows customizing the "zoom center" rather than always scrolling to `_location.top, _location.left` and then having to manually adjust it separately. Having to compute the "scroll adjustment" can be tricky if you don't know exactly how the various position/zoom data from the viewer fit together.
- [ ] an `updateScale` function, that zooms in if the `scaleFactor > 1 || steps > 0`, and zooms out if `scaleFactor < 1 || steps < 0`

The first one is more important, since it centralizes the "adjust scroll after zoom" logic in a single place (the viewer) rather than it being split across the viewer and the wrapper. The second one is just for convenience.

### PR DESCRIPTION
**Pull request by nicolo-ribaudo**
Closes #18076

This PR extends the zoom/scale API to simplify the logic in custom viewer wrappers:
- the first commit provides an `updateScale` function that does both `increaseScale`/`decreaseScale`. Ad you can see from the commit itself, it significantly reduces code duplication both in embedders (see the changes in `web/app.js`) and in `web/pdf_viewer.js` itself
- the second commits moves the "preserve zoom origin when zooming" logic from the app to the viewer, since embedders that want to implement pinch-to-zoom or zoom-around-cursor would need to also implement it (and figuring out exactly what the logic needs to be can be challenging). Part of the "scroll after zoom" logic was already in the viewer, now all of it is roughly in the same place.

Note that this PR increases the net number of lines because it adds a test, but ignoring the new test it would be +56 -95.

@Snuffleupagus In https://github.com/mozilla/pdf.js/pull/18098#issuecomment-2112821398 you said that you probably do not want `updateScale`, but I still uploaded it since it de-duplicates a significant amount of code. What do you think about it? (the second commit could be done independently from the first one)

---

Commits:

> **Unify `increaseScale`/`decreaseScale` logic as `updateScale`**
>
> `updateScale` receives a `drawingDelay`, a `scaleFactor` and/or a number of `steps`.
> If `scaleFactor` is a positive number different from `1` the current scale is multiplied by
> that number. Otherwise, if `steps` if a positive integer the current scale is multiplied by
> `DEFAULT_SCALE_DELTA` `steps` times. Finally, if `steps` is a negative integer, the
> current scale is divided by `DEFAULT_SCALE_DELTA` `abs(steps)` times.

> **Add `origin` parameter to `updateScale`**
>
> This parameter allows defining which point should remain
> fixed while scaling the document. It can be used, for example,
> to implement "zoom around the cursor" or "zoom around
> pinch center".
>
> The logic was previously implemented in `web/app.js`, but
> moving it to the viewer scaling utilities themselves makes it
> easier to implement similar zooming functionalities in
> other embedders.

### CODE REVIEW

File: test/integration/viewer_spec.mjs [hunk 1]
```diff
@@ -17,10 +17,106 @@ import {
...
+  getSpanRectFromText,
   loadAndWait,
 } from "./test_utils.mjs";
 
 describe("PDF viewer", () => {
+  describe("Zoom origin", () => {
+    let pages;
+
+    beforeAll(async () => {
+      pages = await loadAndWait("tracemonkey.pdf", ".textLayer .endOfContent");
```
Comment (Snuffleupagus) [inline test/integration/viewer_spec.mjs:29]: Could we utilize existing viewer functionality to set the desired zoom level, and possibly even the page number, directly here?
```suggestion
      pages = await loadAndWait("tracemonkey.pdf", ".textLayer .endOfContent", "page-width", null, { page: 2 });
```

File: web/pdf_viewer.js [hunk 2]
```diff
@@ -2122,54 +2132,52 @@ class PDFViewer {
...
+    } else if (steps) {
+      const delta = steps > 0 ? DEFAULT_SCALE_DELTA : 1 / DEFAULT_SCALE_DELTA;
+      const round = steps > 0 ? Math.ceil : Math.floor;
+      steps = Math.abs(steps);
       do {
-        newScale =
-          Math.ceil((newScale * DEFAULT_SCALE_DELTA).toFixed(2) * 10) / 10;
-      } while (--steps > 0 && newScale < MAX_SCALE);
+        newScale = round((newScale * delta).toFixed(2) * 10) / 10;
+      } while (--steps > 0 && newScale < MAX_SCALE && newScale > MIN_SCALE);
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:2161]: Given the clamping just below, do we still need this "full" condition here or could it be simplified as follows?
(Obviously it'd lead to a tiny bit of unnecessary computation at the ends of the [MIN_SCALE, MAX_SCALE] range, but I don't know how much that really matters.)
```suggestion
      } while (--steps > 0);
```

Comment (nicolo-ribaudo) [reply web/pdf_viewer.js:2161]: I assume this is not called in a hot path, and that usually the number of steps is low anyway, so yes this should be fine.

File: test/integration/viewer_spec.mjs (hunk 1 above)
Comment (nicolo-ribaudo) [reply test/integration/viewer_spec.mjs]: Thanks for the suggestion, it's much simpler now :)

**Comment by Snuffleupagus** [comment]
> In [#18098 (comment)](https://github.com/mozilla/pdf.js/pull/18098#issuecomment-2112821398) you said that you probably do not want `updateScale`, but I still uploaded it since it de-duplicates a significant amount of code. What do you think about it? 

Looking at a working patch, written by an actual human, I'm happy to retract that statement since it's clear that it avoids a bunch of code duplication :-)

File: web/pdf_viewer.js (hunk 2 above)
Comment (Snuffleupagus) [reply web/pdf_viewer.js]: > I assume this is not called in a hot path, and that usually the number of steps is low anyway,

Correct.

**Comment by nicolo-ribaudo** [comment]
This is the test failing on windows (line 727 is throwing): https://github.com/mozilla/pdf.js/blob/95a7de9f986c449cb949d39af948dbe97f60948e/test/integration/stamp_editor_spec.mjs#L713-L730

It is failing because the `.move()` on line 727 is moving to outside of the viewport. It _seems_ unrelated, because that test doesn't change zoom in any way, however looking at recent CI bot I couldn't find any similar failure 🤔

File: web/pdf_viewer.js [hunk 3]
```diff
@@ -2122,54 +2132,52 @@ class PDFViewer {
...
   /**
-   * Increase the current zoom level one, or more, times.
+   * Changes the current zoom level by the specified amount.
    * @param {ChangeScaleOptions} [options]
    */
-  increaseScale({ drawingDelay, scaleFactor, steps } = {}) {
+  updateScale({ drawingDelay, scaleFactor = null, steps = null, origin }) {
+    if (steps === null && scaleFactor === null) {
+      throw new Error(
+        "Invalid updateScale options: either `scale` or `scaleFactor` must be provided."
```
Comment (Snuffleupagus) [inline web/pdf_viewer.js:2146]:
```suggestion
        "Invalid updateScale options: either `steps` or `scaleFactor` must be provided."
```

**Review by Snuffleupagus — APPROVED** [review]
r=me, with a couple of final comments; thank you!

**Pull request merged**